### ISSUE TITLE
Add Http-Header option dont work on all requests (--http-header) e.g. HLS Segment key URL

### ISSUE THREAD

**Issue by Hoonkr12** [streamlink/streamlink#3754]
## Bug Report

- [x] This is a bug report and I have read the contribution guidelines.
- [ ] I am using the latest development version from the master branch.

### Description

I have to use the Option --http-header to set a header to get my stream working. The Manual for --http-header says "A header to add to each HTTP request." But this isn't true for HLS Key Requests to another URL. 

### Expected / Actual behavior

-http-header is added to all requests and not only to stream url, also to decryptor url

### Reproduction steps / Explicit stream URLs to test

Use streamlink on a hls link where you need a HTTP Header set a see if passes the header for the decryptor url too.

### Log output

HTTP Referer Set, you can see that the m3u8 is downloaded but the key server forbids the requests because there is no Referer set on that request. If i test the decryptor key server link with another tool example my browser and developer tools i can access the key site with the same HTTP Referer Header Set as i need to use for the .m3u8 link:

```

$ streamlink --loglevel debug --http-header "Referer=xxxxxxxxxxxxxxxxxx" https://xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.m3u8 best
[cli][debug] OS:         Linux-5.4.118-1-MANJARO-x86_64-with-glibc2.33
[cli][debug] Python:     3.9.5
[cli][debug] Streamlink: 2.1.1
[cli][debug] Requests(2.25.1), Socks(1.7.1), Websocket(0.59.0)
[cli][debug] Arguments:
[cli][debug]  url=https://xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.m3u8
[cli][debug]  stream=['best']
[cli][debug]  --loglevel=debug
[cli][debug]  --http-header=[('Referer', 'xxxxxxxxxxxxxxxxxxxx/')]
[cli][info] Found matching plugin hls for URL https://xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.m3u8
[plugins.hls][debug] URL=https://xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.m3u8; params={}
[utils.l10n][debug] Language code: de_DE
[cli][info] Available streams: live (worst, best)
[cli][info] Opening stream: live (hls)
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Segments in this playlist are encrypted
[stream.hls][debug] First Sequence: 10176; Last Sequence: 10183
[stream.hls][debug] Start offset: 0; Duration: None; Start Sequence: 10181; End Sequence: None
[stream.hls][debug] Adding segment 10181 to queue
[cli][debug] Pre-buffering 8192 bytes
[stream.hls][debug] Adding segment 10182 to queue
[stream.hls][debug] Adding segment 10183 to queue
[stream.hls][error] Failed to create decryptor: Unable to open URL: https://key.xxxxx.yyyyy/xxxxxx/xxxxxxx (403 Client Error: Forbidden for url: https://key.xxxxx.yyyyy/xxxxxx/xxxxxxx)
[stream.segmented][debug] Closing writer thread
[stream.segmented][debug] Closing worker thread
[cli][error] Try 1/1: Could not open stream <HLSStream('https://exxxxxxxxxxxxxxxxxxxx.m3u8', None)> (No data returned from stream)
error: Could not open stream <HLSStream('https://xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.m3u8', None)>, tried 1 times, exiting
[cli][info] Closing currently open stream...

```

**Comment by bastimeyer** [comment]
> I have to use the Option --http-header to set a header to get my stream working. The Manual for --http-header says "A header to add to each HTTP request." But this isn't true for HLS Key Requests to another URL.

Streamlink uses the same HTTP session for all HTTP requests, and custom HTTP headers get added to the HTTP session before any requests are being made. The HLS stream module also doesn't set or override custom headers in its various HTTP requests, which means any custom headers you set will be included in every HTTP request.

Even though there are no test cases written which check the custom headers in HLS segment key requests, I can confirm in a quick local test on my system that they get set correctly.

**Make sure that you have set the correct headers** and that there is no additional condition to make a successful segment key request.

----

- HTTP session init (used by plugins and stream implementations):
  https://github.com/streamlink/streamlink/blob/5e65d2d9227814ea303300b4dd7269c215fa1865/src/streamlink/session.py#L34
- HTTP session headers update (called from the CLI's `setup_http_session`):
  https://github.com/streamlink/streamlink/blob/5e65d2d9227814ea303300b4dd7269c215fa1865/src/streamlink/session.py#L291-L295
- HTTP requests (global session headers always override specific request headers, if they exist):
  https://github.com/streamlink/streamlink/blob/5e65d2d9227814ea303300b4dd7269c215fa1865/src/streamlink/plugin/api/http_session.py#L120-L131
- HLS segment requests (uses/copies the HLS reader's request params):
  https://github.com/streamlink/streamlink/blob/5e65d2d9227814ea303300b4dd7269c215fa1865/src/streamlink/stream/hls.py#L93-L127
- HLS segment decryption key requests (uses the HLS reader's request params):
  https://github.com/streamlink/streamlink/blob/5e65d2d9227814ea303300b4dd7269c215fa1865/src/streamlink/stream/hls.py#L58-L91
- HLS reader request params (custom params provided by a plugin):
  https://github.com/streamlink/streamlink/blob/5e65d2d9227814ea303300b4dd7269c215fa1865/src/streamlink/stream/hls.py#L341
- custom HLS input plugin:
  https://github.com/streamlink/streamlink/blob/5e65d2d9227814ea303300b4dd7269c215fa1865/src/streamlink/plugins/hls.py#L41-L51

**Comment by Hoonkr12** [comment]
Hi,

huh that was a detailed answer to flush my report and thoughts away. Thank you for that.

I tried to reproduce the issue to debug it further. 
but I couldn’t access the hls key page with wget even with the Headers set…
Tried it in my browser and the only one which I need to set for successful request is the right Referer Header…
Browser is completely set to private browsing and I checked for alle things like cookies and other headers. But only thing was the Referer. Even tried to set the User Agent from my browser in wget. That hasn’t helped.

My Network Knowledge ends on that point. I don’t know if there are some things done to Fingerprint/pin point my bowser or device capabilities for letting me access this HLS segment key page. Handmade DRM to block streamlink?
